pytests/test_debugging.py

```
@@ -0,0 +1,504 @@
+        # breakpoint to any point in the script.
+        self._env.assertGreater(len(response[0]["result"]["locations"]), 0)
+        breakpoint = Breakpoint(self, response[0]["result"])
+        self.breakpoints.append(breakpoint)
+        return breakpoint
+
+    """
+    Removes the breakpoint with the id passed.
+    """
+    def remove_breakpoint(self, breakpoint_id: str):
+        response = self.send_message(DebuggerClientMessage.remove_breakpoint(breakpoint_id))
+        self._env.assertContains("result", response[0])
+        self._env.assertEqual({}, response[0]["result"])
+
+    """
+    Resumes the execution (if it was paused) and returns all the events
+    in response to the resume request.
+    """
+    def resume(self) -> [dict]:
+        self.pause = None
+        return self.send_message(DebuggerClientMessage.DEBUGGER_RESUME)
+
+    """
+    Resumes the execution (if it was paused), waits until the next
+    pause happens and returns all the messages in between.
+    """
+    def resume_and_wait_until_pauses(self) -> [dict]:
+        messages = self.resume()
+        while True:
+            message = self.receive_event()
+            messages.append(message)
+            if "method" in message and message["method"] == "Debugger.paused":
+                self.pause = message
+                break
+        return messages
+
+    """
+    Debugger: step into the stack frame.
+    """
+    def step_into(self) -> [dict]:
+        return self.send_message(DebuggerClientMessage.DEBUGGER_STEP_INTO)
+
+    """
+    Debugger: step out of the current stack frame.
+    """
+    def step_out(self) -> [dict]:
+        return self.send_message(DebuggerClientMessage.DEBUGGER_STEP_OUT)
+
+    """
+    Debugger: step over the next instruction.
+    """
+    def step_over(self) -> [dict]:
+        return self.send_message(DebuggerClientMessage.DEBUGGER_STEP_OVER)
+
+
+def deploy_script(env: Env):
+    env.expect('TFUNCTION', 'DELETE', 'foo')
```

We use `expect` if we also want to check the reply. If we do not care about the reply probably its better to use `cmd` or `execute_command`

---

This requires three more lines to catch the error that the library isn't there. Are you sure you want readability with (in my opinion) excessive try/catch instead of the `expect`, which simply ignores the error (it presumes that the user will check for it), which we don't need? We don't care if the library is there.
I just removed this line for good in favour of using the `REPLACE` flag of the script load instead.

---

So why do we even need to delete it? Why not just use the replace option on the `TFUNCTION LOAD` command?

---

I am already doing it, as mentioned.

---

Ohh great